### PR TITLE
Replace use of 2to3 with six

### LINKED ISSUE #483: TableEditor row_factory insertion not working in Python 3.6
Below is a minimal reproduction of the problem encountered. This works as expected in Python 2.7. When you right click the table and select "add new item" it produces a new row containing a *Person* object with the attributes defined in *row_factory_kw*. In Python 3.6 it inserts an empty uneditable row, and does not appear to modify *people*.  (See screenshot)

```
from traits.api import (List, HasTraits, Unicode)
from traitsui.api import (View, UItem, TableEditor,
    TextEditor, ObjectColumn)

class Person(HasTraits):

    first_name = Unicode

    last_name = Unicode

class People(HasTraits):

    people = List

    traits_view = View(
        UItem('people', editor=TableEditor(
        columns=[
            ObjectColumn(
                name='first_name',
                editor=TextEditor()
            ),
            ObjectColumn(
                name='last_name',
                editor=TextEditor()
            )
        ],
        row_factory=Person,
        row_factory_kw={'first_name':'F','last_name':'L'}
        ))
    )

if __name__ == '__main__':
    People(
        people=[
            Person(first_name='Firstname', last_name='Lastname')
        ]
    ).configure_traits()
```

screenshot from 2018-11-19 10-56-22

### PR DESCRIPTION
**Pull request by corranwebster**
This PR is almost completely automated via `python-modernize`.  Each fixer was applied only if it did anything, and only if what it did looked correct.

The application of the dictionary fixes in the Qt table_editor.py and table_model.py was manually removed as this is an incorrect usage of the fix - see #483 

This PR fixes #483

### CODE REVIEW

**Comment by corranwebster** [comment]
Note for reviewers: since this is largely automated, I think that a check that the process is valid is more important than quality of code.  I am planning a second PR which does manual clean-up of, eg. `six.moves` imports of `range` which aren't really needed.

File: docs/source/tutorials/code_snippets/mpl_figure_editor.py [hunk 1]
```diff
@@ -34,7 +35,7 @@ def _create_canvas(self, parent):
         sizer.Add(mpl_control, 1, wx.LEFT | wx.TOP | wx.GROW)
         toolbar = NavigationToolbar2Wx(mpl_control)
         sizer.Add(toolbar, 0, wx.EXPAND)
-        self.value.canvas.SetMinSize((10,10))
+        self.value.canvas.SetMinSize((10, 10))
```
Comment (mdickinson) [inline docs/source/tutorials/code_snippets/mpl_figure_editor.py]: Ha. Looks like `modernize` sneaks in minor style fixes along the way. (Not objecting; but it's good to know.)

File: docs/source/tutorials/code_snippets/traits_thread.py [hunk 2]
```diff
@@ -1,37 +1,44 @@
...
-    string = String()
+    string = Str()
+
+    view = View(Item('string', show_label=False, springy=True, style='custom'))
 
-    view= View( Item('string',show_label=False, springy=True, style='custom' ))
 
 class CaptureThread(Thread):
     def run(self):
-        self.display.string = 'Camera started\n' + self.display.string  n_img = 0
```
Comment (mdickinson) [inline docs/source/tutorials/code_snippets/traits_thread.py]: Wait, what? How was this even working?

**Comment by mdickinson** [comment]
LGTM based on the description of the method and a statistical sample of the actual changes.

**Comment by mdickinson** [comment]
I'm spotting a couple of uses of the `cmp` builtin function that may need to be fixed.

```
traitsui/wx/image_slice.py:283:39: F821 undefined name 'cmp'
traitsui/wx/image_slice.py:312:39: F821 undefined name 'cmp'
```

File: traitsui/wx/check_list_editor.py [hunk 3]
```diff
@@ -120,7 +123,7 @@ def list_updated(self, values):
                     logger.warn('Unable to remove non-current value [%s] from '
                                 'values %s', cur_value[i], values)
         if modified:
-            if isinstance(self.value, basestring):
+            if isinstance(self.value, six.string_types):
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply traitsui/wx/check_list_editor.py]: I think that this is right in most cases - if we are using `basestring` then we want both `str` and `unicode` under Python 2, and for Python 3 it is just `str`.

I don't think that there are any places where we _really_ want `bytes` as well in Python 3.

**Comment by corranwebster** [comment]
Good catch on the `cmp` - although it looks like they weren't being fixed anyway as there are no 2to3 fixers for `cmp`.  Presumably because there is no general way to automate a replacement.  So they would be bugs in the current code base anyway.

File: traitsui/wx/check_list_editor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdickinson) [reply traitsui/wx/check_list_editor.py]: > I don't think that there are any places where we _really_ want `bytes` as well in Python 3.

Yes; I was mostly thinking about the opposite: there are probably bits of Python 2 code where we're currently accepting both bytestrings and Unicode, and (on general principles) should really only be accepting Unicode. But that's a semantic change in the working Python 2 code, so very much off-topic for this PR. There's a small advantage to having the Python 2 and Python 3 codebases being aligned in what they accept as far as possible, but it probably doesn't outweigh the risk of breaking things in Python 2 land.

**Comment by mdickinson** [comment]
FWIW, the occurrences of `cmp` looked trivial to replace with a `key`.

**Pull request merged**